### DISCUSSION TITLE
Immutability in the EDC

### DISCUSSION THREAD

**Discussion by arnoweiss** [eclipse-edc/Connector#3664] [Q&A] [answered]
I'd like to understand how the EDC views the topic of immutable objects in the Management API.

PUT and DELETE operations are available on the `/v3/assets/{{id}}`, `/v2/policydefinitions/{{id}}` and `/v2/contractdefinitions/{{id}}` endpoints. However, if the resources can be changed only under some circumstances. Here's my observations:

- DELETE /v3/assets/{{id}}
  - if there is a contract agreement associated with the asset in question: HTTP 409. The asset remains unchanged.
  - if there is no contract agreement: HTTP 204. The asset is removed.
- PUT /v3/assets/{{id}}
  - if there is a contract agreement associated with the asset in question: HTTP 200. **The asset remains unchanged**.
  - if there is no contract agreement: HTTP 200. The asset is changed.

As contract agreements cannot be deleted, it is impossible to alter an asset after an associated contract agreement was made. This is reflected by the APIs behavior on DELETE but not on PUT where the return code signals that a change to the resource was applied when it was in fact not.

Let us suppose a scenario where a backend service was migrated from one URL to another. This would have to be reflected in the HttpDataAddress's baseUrl. However, this scenario can't be executed with the PUT /v3/assets/{{id}} request currently. So my question is:

What is the intended behavior for the EDC wrt mutability of assets?

**Comment by paullatzelsperger** [comment, marked as answer]
Immutability of objects is highly influenced by business rules, and possibly even legal requirements. Whether it's acceptable to change an asset, that's referenced in a contract, may vary from jurisdiction to jurisdiction. EDC can't make assumptions about that with any degree of certainty. We do, however, enforce mutability rules where they are technically necessary, for example _deleting_ an asset that is referenced by a contract would leave us with an inconsistent state thus we disallow it. 

Thus, changing a contracted asset may be "unlawful", or against business rules in a particular use case, but those must be enforced elsewhere. 

So a `PUT` on an asset should indeed change it, regardless whether its used in a contract or not. If a 200 is returned, but the asset remains unchanged, that could be a bug. In that case I would ask to reproduce with the latest EDC code base and file a proper bug report.

**Reply by arnoweiss** [reply]
Thanks for the explanation! The behavior I was describing appears to have been fixed sometime between eclipse-edc 0.2.1 and 0.4.1.
